addon/inject.ts

```diff
@@ -4,32 +4,91 @@ import { getOwner } from '@ember/application';
 import { Registry } from './index';
 import Provider from './provider';
 
-const HAVE_MODIFIED = new WeakSet<EmberObject>();
-const PROVIDERS = new WeakMap<EmberObject, Array<Provider>>();
+const INJECTED_PROVIDERS = new WeakMap<EmberObject, Provider[]>();
 
-export default function inject<K extends keyof Registry>(identifier: K) {
+interface IProviderKlass extends Function {
+  create: () => Provider;
+}
+
+interface IEmberObjectWithParentView extends EmberObject {
+  parentView?: IEmberObjectWithParentView;
+}
+
+function findParentProvider(
+  component: IEmberObjectWithParentView,
+  ProviderKlass: IProviderKlass
+): Provider | undefined {
+  if (component.parentView) {
+    const parentProviders = INJECTED_PROVIDERS.get(component.parentView);
+    let provider;
+
+    if (parentProviders) {
+      provider = parentProviders.find(p => p instanceof ProviderKlass);
+    }
+
+    return provider
+      ? provider
+      : findParentProvider(component.parentView, ProviderKlass);
+  }
+
+  return undefined;
+}
+
+/**
+ * Inject a Provider into a host object
+ *
+ * @param identifier the Provider's identifier on the registry
+ * @param options
+ */
+export default function inject<K extends keyof Registry>(
+  identifier: K,
+  { inherit = false } = {}
+) {
   return computed<Registry[K]>(function(this: EmberObject) {
     const owner = getOwner(this);
-    const provider = owner.lookup(`provider:${identifier}`);
+    const ProviderKlass: IProviderKlass = owner.lookup(
+      `provider:${identifier}`
+    );
 
-    // Modify the host class to also destroy providers
-    if (!HAVE_MODIFIED.has(this)) {
+    let provider: Provider;
+    let isOwner = true;
+
+    if (inherit) {
+      const possibleProvider = findParentProvider(this, ProviderKlass);
+
+      if (!possibleProvider) {
+        provider = ProviderKlass.create();
+      } else {
+        isOwner = false;
+        provider = possibleProvider;
+      }
+    } else {
+      provider = ProviderKlass.create();
```

Comment:
I _think_ this might end up breaking `getOwner(this)` for a Provider (in case you wanted to inject a service into a provider or something). 

You can leave [this](https://github.com/alexlafroscia/ember-provider/pull/4/files#diff-d5cbf8bfcd89d878177f9e9b7c02e42aR5) as `singleton: false` and use: 

```js
const ProviderKlass = owner.lookupFactory(`provider:${identifier}`); 
```

And then this line would become: 

```suggestion
      provider = owner.lookup(`provider:${identifier}`);
```

Reply:
Ah, good catch!

You can also manually provide the owner injection to `create`:

```javascript
const owner = getOwner(this);
provider = ProviderKlass.create(owner.ownerInjection());
```

I think I like your approach better, though!

Reply:
Nice! I did not know about `owner.ownerInjection()` — neat!

Reply:
Yeah, I discovered that when writing the tests for this addon -- that's how I create fake objects in different scenarios that still have access to the `owner`

Reply:
Hmm, it doesn't seem like `lookupFactory` is a method... was it `factoryFor`?

Reply:
Whoops — yes, it's likely that. I was improvising and sometimes my fingers land on the wrong keys. :-P

Reply:
Using `factoryFor` bring the `instanceof` check that I'm doing -- I'm just going to add the `owner.ownerInjetion()` bit and see if that helps